### PR TITLE
Ensure PyTorch tensors support CUDA when using `qml.ctrl`

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** A recent [user-found bug](https://discuss.pennylane.ai/t/qml-ctrl-operation-on-gpu/2836/5) indicated that the `qml.ctrl` operator and `qml.RY` operator did not correctly allow parameters living on a CUDA-device to work in a comparable manner with `qml.CRY`. This was due to certain operations creating host-side tensors which were never moved back to the GPU during creation. 

**Description of the Change:** This PR ensures that incoming tensors and intermediates as part of the `qml.ctrl` and single-parametric gate pipeline will be moved to the GPU if a parameter tensor lives there. A simplified test based on the user-supplied example was also added to validate this against GPU devices in future.

**Benefits:** Fixes user-reported bug.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -225,6 +225,9 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Fixes a bug where `qml.ctrl` for parametric gates were incompatible with PyTorch tensors on the GPU.
+  [(#3984)](https://github.com/PennyLaneAI/pennylane/pull/4002)
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:229]:
```suggestion
  [(#4002)](https://github.com/PennyLaneAI/pennylane/pull/4002)
```

**Comment by mlxd** [comment]
> looks good! I did a quick spot-check on single/multi-dispatch to see if there are any other functions that might need this, and I think not but I may have missed something. Now that I'm typing this, I'm worried that `cast_like` (from my suspected modules: found 30 times in transforms/, 8 times in ops/op_math/) will cause this issue elsewhere. I'm not sure where, or how to confirm it... can you/another reviewer confirm or deny my fears? I don't have a GPU smile
> 
> I'm going to approve because the existing changes look good, and we can look into the above in a later PR if there's a rush on this one, but I'm personally good to slow down and address it all in this PR.

Thanks @timmysilv 

I agree that it is very likely this issue will crop up in other parts of the code-base. Maybe merging this to get the fix for the next release is a good first call, and planning to address more extensive tests to catch other such cases later would make sense to me. Happy to be swayed otherwise though.

**Comment by mlxd** [comment]
It is also possible the the `cast_like/convert_like` cascade could be replaced with a single appropriately structured `cast_like`, but I'd consider that only if it made sense to do so.

**Comment by timmysilv** [comment]
True. This is not the first time [someone has wondered that](https://app.shortcut.com/xanaduai/story/33770/add-custom-conversion-method-to-math-module-that-acts-like-convert-like-or-cast-like). Maybe it's time we re-visit this?

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (mlxd) [reply doc/releases/changelog-dev.md]: Ooops, good catch.

**Review by mudit2812 — APPROVED** [review]
Thanks Lee!

**Comment by albi3ro** [comment]
[sc-37211]

**Pull request merged**